### ISSUE TITLE
Memory leak with view controllers creation/removal

### ISSUE THREAD

**Issue by mirkonasato** [robovm/robovm#749]
Given a simple app with a navigation controller repeatedly pushing and popping a view controller, the Swift version shows the following memory usage with the Allocations instrument

![swiftleaker-allocations](https://cloud.githubusercontent.com/assets/133967/6096239/b36cde88-af7e-11e4-874f-800c6724b7f9.png)

while the RoboVM version shows a constantly increasing memory usage

![roboleaker-allocations](https://cloud.githubusercontent.com/assets/133967/6096241/d04792fa-af7e-11e4-8f98-9d176562b62f.png)

Code available at https://github.com/mirkonasato/robovm-issue-749 - graphs created by clicking ~30 times on the Start then Close buttons at ~1 second intervals.

This is with 1.0.0-beta-03.

**Comment by mirkonasato** [comment]
Explicitly calling dispose() on the view controller after popping it is a workaround.

**Comment by badlogic** [comment]
Modified the example, `OtherViewController` is now looking like this:

```java
package example;

import org.robovm.apple.uikit.UINavigationController;
import org.robovm.apple.uikit.UIViewController;

public class OtherViewController extends UIViewController {

    @Override
    public void viewDidLoad() {          
    }    

	@Override
	public void viewDidAppear(boolean animated) {
		UINavigationController navigationController = getNavigationController();
        navigationController.popViewController(false);	
        this.dispose();
	}
}
```
Doing a `leaks` before i click start in the `HomeViewController` gives me https://gist.github.com/badlogic/3350046940ebd3b9e686. Doing a `leaks` after i clicked the start button gives me https://gist.github.com/badlogic/93235b25d8252e4768de. 

The amount of leaking memory stays constant. That means we retain memory via reference chains. Question is whether it's on the Java or native side.

**Comment by badlogic** [comment]
Instruments shows us an increase in memory, counting memory allocated via the GC, but doesn't show us individual objects allocated via the GC, as those are taken from memory blocks controlled by the GC. Trying to figure out if `OtherViewController` instances get collected.

1. Build RoboLeaker in debug mode
2. Fire it up on the sim, attach XCode
3. Pause, and do the following:

```
(lldb) br set -f OtherViewController.java -l 10
Breakpoint 1: where = RoboLeaker`[J]example.OtherViewController.viewDidLoad()V + 73 at OtherViewController.java:10, address = 0x0029a5f9
(lldb) c
Process 3209 resuming
(lldb) fr v
(void *) __$env = 0x11ea4fc8
(void *) __$this = 0x1364bac8
(lldb) memory read 0x1364bac8
0x1364bac8: e0 2a 38 13 00 00 00 00 70 bf d0 11 00 00 00 00  .*8.....p.......
0x1364bad8: 80 cd 68 13 01 00 00 00 00 00 00 00 00 00 00 00  ..h.............
(lldb) c
Process 3209 resuming
(lldb) memory read 0x1364bac8
0x1364bac8: e0 2a 38 13 00 00 00 00 70 bf d0 11 00 00 00 00  .*8.....p.......
0x1364bad8: 80 cd 68 13 01 00 00 00 00 00 00 00 00 00 00 00  ..h.............
```
The GC would null out a GC'ed object, but it appears that's not happening. This should mean that the `OtherViewController` has not been collected.

Trying to trace references to that object next.

**Comment by badlogic** [comment]
`memory.c` has head dumping functionality. The `VM` Java side has a method to trigger.

**Comment by badlogic** [comment]
Instrumented ObjectOwnershipHelper. Here's what happens retain wise when we create, push and pop the `OtherViewController`:

```
Retained instance of example.HomeViewController at 0x12412820, retain count: 3
Retained instance of example.HomeViewController at 0x12412820, retain count: 4
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 2
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 3
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 4
Retained instance of example.HomeViewController at 0x12412820, retain count: 5
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 5
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 6
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 7
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 8
0
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 9
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 10
Retained instance of example.HomeViewController at 0x12412820, retain count: 6
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 11
Released instance of example.HomeViewController at 0x12412820, retain count: 5
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 10
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 11
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 12
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 11
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 12
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 13
Retained instance of example.HomeViewController at 0x12412820, retain count: 6
Released instance of example.HomeViewController at 0x12412820, retain count: 5
Retained instance of example.HomeViewController at 0x12412820, retain count: 6
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 14
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 15
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 16
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 17
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 16
Retained instance of example.HomeViewController at 0x12412820, retain count: 7
Retained instance of example.HomeViewController at 0x12412820, retain count: 8
Retained instance of example.HomeViewController at 0x12412820, retain count: 9
Retained instance of example.HomeViewController at 0x12412820, retain count: 10
Retained instance of example.HomeViewController at 0x12412820, retain count: 11
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 17
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 18
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 17
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 16
Retained instance of example.HomeViewController at 0x12412820, retain count: 12
Retained instance of example.OtherViewController at 0x1a52a4e0, retain count: 17
Released instance of example.HomeViewController at 0x12412820, retain count: 11
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 16
Released instance of example.HomeViewController at 0x12412820, retain count: 10
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 15
Retained instance of example.HomeViewController at 0x12412820, retain count: 11
Retained instance of example.HomeViewController at 0x12412820, retain count: 12
Retained instance of example.HomeViewController at 0x12412820, retain count: 13
Retained instance of example.HomeViewController at 0x12412820, retain count: 14
Retained instance of example.HomeViewController at 0x12412820, retain count: 15
Released instance of example.HomeViewController at 0x12412820, retain count: 14
Retained instance of example.HomeViewController at 0x12412820, retain count: 15
Retained instance of example.HomeViewController at 0x12412820, retain count: 16
Released instance of example.HomeViewController at 0x12412820, retain count: 15
Retained instance of example.HomeViewController at 0x12412820, retain count: 16
Retained instance of example.HomeViewController at 0x12412820, retain count: 17
Retained instance of example.HomeViewController at 0x12412820, retain count: 18
Released instance of example.HomeViewController at 0x12412820, retain count: 17
Retained instance of example.HomeViewController at 0x12412820, retain count: 18
Released instance of example.HomeViewController at 0x12412820, retain count: 17
Released instance of example.HomeViewController at 0x12412820, retain count: 16
Released instance of example.HomeViewController at 0x12412820, retain count: 15
Released instance of example.HomeViewController at 0x12412820, retain count: 14
Released instance of example.HomeViewController at 0x12412820, retain count: 13
Released instance of example.HomeViewController at 0x12412820, retain count: 12
Released instance of example.HomeViewController at 0x12412820, retain count: 11
Released instance of example.HomeViewController at 0x12412820, retain count: 10
Released instance of example.HomeViewController at 0x12412820, retain count: 9
Released instance of example.HomeViewController at 0x12412820, retain count: 8
Released instance of example.HomeViewController at 0x12412820, retain count: 7
Released instance of example.HomeViewController at 0x12412820, retain count: 6
Released instance of example.HomeViewController at 0x12412820, retain count: 5
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 14
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 13
Released instance of example.HomeViewController at 0x12412820, retain count: 4
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 12
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 11
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 10
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 9
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 8
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 7
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 6
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 5
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 4
Released instance of example.HomeViewController at 0x12412820, retain count: 3
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 3
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 2
Released instance of example.OtherViewController at 0x1a52a4e0, retain count: 1
Released instance of example.HomeViewController at 0x12412820, retain count: 2
```

As you can see, the `OtherViewController` has a retain count of 1 after the pop. That is due to the `alloc` call when instantiating the `OtherViewController`:

![http://sht.tl/Wh1Z08](http://sht.tl/Wh1Z08)

That retain must be countered with a release by us, which should happen in the GC finalizer of that object.

**Comment by badlogic** [comment]
The finalizer can't  happen because the retain count is never 1, and hence the object sticks around in custom objects, so it's uncollectable.

**Comment by badlogic** [comment]
Testing, using the original RoboLeaker code, modified as follows:

1. Calling `ObjCObject.logRetainRelease(true);` in `RoboLeaker#main`
2. Remove `closeButton.addOnTouchUpInsideListener((control, event) -> close());` from `OtherViewController` and call `close` at the end of `viewDidLoad` directly. This eliminates the associated object leak we still have, e.g. the UIButton listener wrapper would retain the view controller
3. Add an instance field `byte[] chunk  = new byte[512*1026]`, this way the view controller is bigger in size and it's easier to gauge the leak as the OS does allocate other tiny stuff as well.
3. A button in `HomeViewController` to manually invoke the GC:

```java
package example;

import org.robovm.apple.coregraphics.CGRect;
import org.robovm.apple.uikit.UIButton;
import org.robovm.apple.uikit.UIButtonType;
import org.robovm.apple.uikit.UIControlState;
import org.robovm.apple.uikit.UINavigationController;
import org.robovm.apple.uikit.UIViewController;

public class HomeViewController extends UIViewController {

    @Override
    public void viewDidLoad() {
        super.viewDidLoad();
        UIButton startButton = UIButton.create(UIButtonType.System);
        startButton.setFrame(new CGRect(115.0f, 121.0f, 91.0f, 37.0f));
        startButton.setTitle("Start", UIControlState.Normal);
        startButton.addOnTouchUpInsideListener((control, event) -> showOtherViewController());
        getView().addSubview(startButton);
        
        UIButton gcButton = UIButton.create(UIButtonType.System);
        gcButton.setFrame(new CGRect(115.0f, 160.0f, 91.0f, 37.0f));
        gcButton.setTitle("GC", UIControlState.Normal);
        gcButton.addOnTouchUpInsideListener((control, event) -> {
        	for(int i = 0; i < 100; i++) {
        		System.gc();
        	}
        });
        getView().addSubview(gcButton);
    }

    private void showOtherViewController() {
        UINavigationController navigationController = getNavigationController();
        OtherViewController otherViewController = new OtherViewController();
        navigationController.pushViewController(otherViewController, false);
    }
}
```

### Test Procedure
1. deploy the app to the sim, in debug mode
2. Attach Instruments once the HomeViewController is visible
3. Switch to and from the OtherViewController 20 times
4. Let the app settle on the HomeViewController, ARC will kill unrelated objects, e.g. status bar, dispatch blocks etc.
5. Inspect persisted bytes in Instruments, as well as checking for retained instances of `OtherViewController` (use a filter of `robovm` in Instruments).
6. Invoke the GC manually via button in HomeViewController
7. Inspect persisted bytes in Instruments, as well as checking for retained instances of `OtherViewController` (use a filter of `robovm` in Instruments).

### Test 1: No registration in ObjectOwnershipHelper
The first test is without registration in `ObjectOwnershipHelper`. This means we don't hook retain/release, and we don't keep a reference of the view controller on the Java side.

Before GC (step 5):
![http://sht.tl/sygHbV](http://sht.tl/sygHbV)
![http://sht.tl/FCioKx](http://sht.tl/FCioKx)

The app has settled at around 190kb. There are `OtherViewController` instances with a retain count of 1

After GC (step 7):
![http://sht.tl/QJ2pmJ](http://sht.tl/QJ2pmJ)
![http://sht.tl/Dq2Glh](http://sht.tl/Dq2Glh)

The GC invoke actually added persisted bytes, probably allocated more pages. The `OtherViewController` instances and contained, retained objects are still there.

### Test 2: Registration in ObjectOwnershipHelper, removing from `CUSTOM_OBJECTS` on a retain count of 1
In this test, we register the custom `OtherViewController` with `ObjectOwnershipHelper`, but we only remove the object in `ObjectOwnershipOwner#release` if the retain count is 1. This is wrong, we need to remove the object from the map at a retain count of 2, as the count indicates the retain count BEFORE the current retain is complete. If we don't do this, the GC will never be able to collect and hence release the object as it will always be referenced by CUSTOM_OBJECTS.

Before GC (step 5)
![http://sht.tl/Q3y06v](http://sht.tl/Q3y06v)
![http://sht.tl/7pFhuk](http://sht.tl/7pFhuk)

The app settled at around 190kb, There are `OtherViewController` instances with a retain count of 1.

After GC (step 7)
![http://sht.tl/xEcaq2](http://sht.tl/xEcaq2)
![http://sht.tl/rjlVCp](http://sht.tl/rjlVCp)

`OtherViewController`s remain unreleased, memory increased again slightly after the GC. The retain/release logging tells us that all `OtherViewController`s have a retain count of 1, so does instruments.

### Test 3: Registration in ObjectOwnershipHelper, removal from `CUSTOM_OBJECTS` on a retain count of 2

Before GC (step 5):
![http://sht.tl/KNezMn](http://sht.tl/KNezMn)
![http://sht.tl/KNezMn](http://sht.tl/vRB1oC)

App settled at around 80kb, it could already collect a few of the `OtherViewController` instances.

After GC (step 7):
![

**Comment by badlogic** [comment]
Added a fix for custom objects not being released properly, both on the Java and native side. Setup a test based on your example, with modifications. See https://github.com/badlogic/robovm-issue-749/commits/master

**Note that i disabled the touch listener on the button. We still have a leak with associated objects. This only shows that custom objects do not leak with the fix in #825.**

#### Test Procedure
1. Compile app in release mode, start on sim
2. Wait for home view controller to show up, attach instruments via the allocation profile to the running app
3. Click on start like crazy, for say 30 seconds
5. Stop clicking, let the app settle, check native memory in instruments, persisted bytes as well as #persisted of OtherViewController. Check total and used bytes in Eclipse console for Java heap memory.
6. Press GC button
7. Check native memory in instruments, persisted bytes as well as #persisted of OtherViewController. Check total and used bytes in Eclipse console for Java heap memory.

#### Test 1: No registration with ObjectOwnershipHelper
Before GC
![http://sht.tl/fJVZpS](http://sht.tl/fJVZpS)
Total: 5.3984375
Used: 3.1484375

After GC
![http://sht.tl/VEj9Ng](http://sht.tl/VEj9Ng)
Total: 2.7148438
Used: 2.7148438

Some `OtherViewController`s are collected after a while when the GC pressure is high enough, calling release on the underlying native object. When we call GC manually, all native objects disappear, both the native `OtherViewController`s and any objects they retained. The Java heap also shrinks again.

However, the GC might decide to kill the `OtherViewController` because it's not referenced from the Java-side or resident in any GC roots. We thus introduced `ObjectOwnershipHelper`.

#### Test 2: Registration with ObjectOwnershipHelper, before #825
Before GC
![http://sht.tl/cchiZV](http://sht.tl/cchiZV)
Total: 5.3671875
Used: 4.140625

After GC
![http://sht.tl/pKUU5P](http://sht.tl/pKUU5P)
Total: 2.8242188
Used: 2.8242188

No `OtherViewController`s can be collected, because they are referenced from `CUSTOM_OBJECTS`. This means they can also not be released on the native side. After the GC we freed up some Java heap (most likely the strings we create during logging), but the native side stays the same. You will also see that no release count logged will ever be 0.

#### Test 3: Registration with ObjectOwnershipHelper, after #825
Before GC
![http://sht.tl/3mm5mD](http://sht.tl/3mm5mD)
Total: 4.0
Used: 3.9023438

After GC
![http://sht.tl/Acemdk](http://sht.tl/Acemdk)
Total: 2.2460938
Used: 2.2460938

The GC can kill some of the `OtherViewController`s on the Java heap side because we now properly remove them from `CUSTOM_OBJECTS` on retain. This also means we release the native parts of these controllers properly. Once we GC manually, the rest of the controllers are gone, both on the Java and on the native side.

#### Conclusion
So, #825 fixes custom objects leaking. `ObjectOwnershipHelper` was introduced after OP's issue report. The cause of OP's issue are associated objects (touch listeners, etc.). Without the touch listener in OP's original report, we'd end up with the results from test 1 above.

I'll fix up associated objects in #825 and do another round of tests.

**Comment by badlogic** [comment]
Simplifying this a little. I created a console project and added the objc runtime to it. Here's the sample:

```java
import org.robovm.apple.foundation.NSObject;
import org.robovm.objc.ObjCObject;

public class AssociatedObject {
	private static class CustomObject extends NSObject { }
		
	private static void doWork() {
		CustomObject obj = new CustomObject();
		obj.setAssociatedObject("assosiacted", "My Value");
		System.out.println("Before leaving scope");
	}
	
	public static void main(String[] args) {
		ObjCObject.logRetainRelease(true);
		doWork();
		
		// this will call the finalizer on the CustomObject
		// and release it, which should also trigger the
		// release of the associated object. Check console
		System.out.println("Before GC");
		for(int i = 0; i < 1000; i++) {
			System.gc();
		}
		System.out.println("After GC");
	}
}
```

I create a custom object (so i can log its release/retain, can't do it for native object). Next i set an associated object on it, in this case a String. There's no cycle.

I then:
1. Set a breakpoint before calling `doWork` and on `System.out.println("Before GC")` and on `System.out.println("After GC")`
2. Once i hit the first breakpoint, i attach Instruments
3. I continue execution on the Java side until it hits the second breakpoint

I get this
![http://sht.tl/8ZeZpQ](http://sht.tl/8ZeZpQ)

What we expected. Instruments shows our custom object (its native peer), and a RoboVMReleaseListener which wraps the associated object. We can also see, that the RoboVMReleaseListener has already been released once. That's what we do manually in AssociatedObjectHelper. The listener now has a retain count of 1.

I continue again until i hit the third breakpoint.
![http://sht.tl/JsDLdu](http://sht.tl/JsDLdu)

The GC collects the out-of-scope custom object, which triggers a release, which triggers a release on the associated object. Per the retain/release log, both our custom object and our associated object have been released and deallocated. Instrument also tells us that the native peers are gone.

This means `AssociatedObjectHelper` works as intended as long as there are no reference cycles on the Java-side. Setting the associated object does not incur a retain of the object on which its set.

The whole concept breaks down if we have a cyclic reference on the Java side between the object and the associated object set on that object. A cyclic reference on the Java side OK as long as none of the referents are reachable from GC roots (think disconnected component in the object graph). 

However, in the case of listeners on `UIControl` we have such a GC reachable referent in form of `ASSOCIATED_OBJECTS`. If someone assigns a listener that references the UIControl directly or indirectly, the reference cycle is reachable from `ASSOCIATED_OBJECTS`. We can not prevent users from doing this. In fact it's rather "easy" to do with anonymous inner classes/lambdas.

I think we need to rethink the way we used associated objects.

**Comment by badlogic** [comment]
Updated sample:

```java
import org.robovm.apple.foundation.NSObject;
import org.robovm.objc.ObjCObject;

public class AssociatedObject {
	private static class CustomObject extends NSObject { }
		
	private static void doWork() {
		CustomObject obj = new CustomObject();
		obj.setAssociatedObject("assosiacted", "String");
		System.out.println("Before leaving scope");
	}
	
	private static void lambda() {
		CustomObject obj = new CustomObject();
		Runnable lambda =  () -> System.out.println("Test");
		obj.setAssociatedObject("assosiacted", lambda);
		System.out.println("Before leaving scope");
	}
	
	public static void main(String[] args) {
		ObjCObject.logRetainRelease(true);
		lambda();
		
		// this will call the finalizer on the CustomObject
		// and release it, which should also trigger the
		// release of the associated object. Check console
		System.out.println("Before GC");
		for(int i = 0; i < 1000; i++) {
			new String(new byte[100]);
			System.gc();
		}
		System.out.println("After GC");
	}
}
```
The lambda doesn't capture the object, release/GC works as intended.

**Comment by badlogic** [comment]
We need to break the cyclic reference on the Java side when to native/custom object is released the last time from the native side. At that point in time the custom/native object has a release count of 2 (1 for the last native retain, and one for our retain in alloc/init). 

For custom objects we could clean up the associated object and the map entry in ASSOCIATED_OBJECTS in the swizzled retain when retain count is 2.

For native objects like UIButton i'm not sure what the best solution would be. We could swizzle that retain as well i guess, though that has a slight performance impact.

I can think of one scenario where this fails: Imagine we have a UIButton, add it as a sub view, and keep a reference to it on the Java side. Eventually, the button gets removed from the subview and any listeners get cleaned up on retain count == 2. Now the button has a retain count of one, and all listeners have beem whiped out. If we add that button to another view, the listeners aren't there anymore. I'm not sure why anyone would do this, and i'm also not sure if one is allowed to reuse a UIControl like that.

There must be a better way to break the cycle.

**Comment by badlogic** [comment]
I can not think of a way to break the reference cycle in the artifical example above. However, for `RoboLeaker` it's easy to break the cycle between the listener and the `OtherViewController`. Here's the modified `OtherViewController` that neither leaks custom objects (the controller), not associated objects (the touch listener), when #825 is applied:

```java
package example;

import org.robovm.apple.coregraphics.CGRect;
import org.robovm.apple.uikit.UIButton;
import org.robovm.apple.uikit.UIButtonType;
import org.robovm.apple.uikit.UIControl;
import org.robovm.apple.uikit.UIControl.OnTouchUpInsideListener;
import org.robovm.apple.uikit.UIControlState;
import org.robovm.apple.uikit.UIEvent;
import org.robovm.apple.uikit.UINavigationController;
import org.robovm.apple.uikit.UIViewController;

public class OtherViewController extends UIViewController {
	
    @Override
    public void viewDidLoad() {
        super.viewDidLoad();
        UIButton closeButton = UIButton.create(UIButtonType.System);
        closeButton.setFrame(new CGRect(115.0f, 121.0f, 91.0f, 37.0f));
        closeButton.setTitle("Close", UIControlState.Normal);
        closeButton.addOnTouchUpInsideListener(new TouchListener());
        getView().addSubview(closeButton);
    }
    
    private static class TouchListener implements OnTouchUpInsideListener {
		@Override
		public void onTouchUpInside(UIControl control, UIEvent event) {			
	        UINavigationController navigationController = (UINavigationController)control.getWindow().getRootViewController();
	        navigationController.popViewController(false);
		}
    }
}
```

This touch listener doesn't reference the OtherViewController, which can thus be collected by the GC, which in turn will trigger a release and GC of the touch listener and the associated object. I did testing via Instruments and none of our objects are leaked. 

However, it appears that CoreAnimation is leaking memory. I did some generation analysis:
![http://sht.tl/Quqb4A](http://sht.tl/Quqb4A)

Most of the additional memory between generations comes from CoreAnimation.

**Comment by badlogic** [comment]
Did the same analysis on a device, similar results. It appears that iOS caches some things and eventually reuses them or throws them away. I assume that's based on memory pressure, of which there isn't a lot in the RoboLeaker example.

![http://sht.tl/7YhZBy](http://sht.tl/7YhZBy)

**Comment by badlogic** [comment]
I tested the SwiftLeaker version of this app on the device, it exhibits the exact same behaviour as our RoboLeaker version with #825 applied.

![http://sht.tl/O3z9YV](http://sht.tl/O3z9YV)

So my theory may be correct about iOS caching and reusing objects.

*The conclusion is as follows:*

@ntherning applying #825 is required to close memory leaks of custom objects. We can merge that.

Reference cycles on the Java side between controllers, controls and listeners will lead to leaks. If you set something like a touch listener, don't reference the control or the view controller inside that listener. Anonymous inner classes and lambdas may capture things without you noticing.

I'm not giving up on breaking the cycles between objects and their associated objects just yet.

**Comment by badlogic** [comment]
The retain counts shown by Instruments are incorrect:

![http://sht.tl/G8DiVA](http://sht.tl/G8DiVA)

Related: i'm not quite sure why the button would have a retain count of 2 after `UIButton#create`

**Comment by badlogic** [comment]
All our `ListenerWrapper`s are custom objects. Putting them into `ASSOCIATED_OBJECTS` is a waste, they are also stored in `CUSTOM_OBJECT`s and will hence survive collection of their parent object, e.g. `UIButton`.

**Comment by badlogic** [comment]
Turns out listener wrappers are not registered, and hence not in `CUSTOM_OBJECTS`.

**Comment by badlogic** [comment]
Here's a pattern that makes sure your capturing listeners don't introduce a reference cyle:

1. Layout your view in `UIViewController#viewDidLoad`
2. Add your listeners in `UIViewController#viewWillAppear`
3. Remove your listeners in `UIViewController#viewWillDisappear`

For the `OtherViewController` of the RoboLeaker sample this would look as follows:

```java
package example;

import org.robovm.apple.coregraphics.CGRect;
import org.robovm.apple.uikit.UIButton;
import org.robovm.apple.uikit.UIButtonType;
import org.robovm.apple.uikit.UIControl.OnTouchUpInsideListener;
import org.robovm.apple.uikit.UIControlState;
import org.robovm.apple.uikit.UIViewController;

public class OtherViewController extends UIViewController {
	OnTouchUpInsideListener listener = null;
	UIButton closeButton = null;

	@Override
	public void viewDidLoad() {
		super.viewDidLoad();
		closeButton = UIButton.create(UIButtonType.System);
		closeButton.setFrame(new CGRect(115.0f, 121.0f, 91.0f, 37.0f));
		closeButton.setTitle("Close", UIControlState.Normal);
		getView().addSubview(closeButton);
	}

	@Override
	public void viewWillAppear(boolean animated) {
		closeButton.addOnTouchUpInsideListener(listener = (control, event) -> {
			getNavigationController().popViewController(false);
		});
	}

	@Override
	public void viewWillDisappear(boolean animated) {
		super.viewWillDisappear(animated);
		closeButton.removeListener(listener);
	}
}
```

The upside is, that you separate your layouting code from your event management code. The downside is that you now have to keep references to your ui controls and listeners around so you can remove the listeners in `viewWillDisappear`. I guess we could add a `UIControl#removeAllListeners` method (and do the same for other classes that employ this pattern). That way you'd only need to keep track of your ui controls, which you probably already have to do anyways.

**Comment by ntherning** [comment]
@badlogic You might already have covered it in the previous comments but I'm too lazy - what does the cycle typically look like? I think it would help to have a drawing of some sort to illustrate it.

**Comment by badlogic** [comment]
Given this `UIViewController`:
```java
public class OtherViewController extends UIViewController {
	@Override
	public void viewDidLoad() {
		super.viewDidLoad();
		UIButton closeButton = UIButton.create(UIButtonType.System);
		closeButton.setFrame(new CGRect(115.0f, 121.0f, 91.0f, 37.0f));
		closeButton.setTitle("Close", UIControlState.Normal);
		closeButton.addOnTouchUpInsideListener((control, event) -> {
			getNavigationController().popViewController(false);
		});
		getView().addSubview(closeButton);
	}
}
```

Let's assume that we pushed the controller to a navigation controller, and the controller went out of scope. We end up with this object graph:

![https://s3.amazonaws.com/pushbullet-uploads/ujEbQh7xysS-PaQBCYlDl5pfMlLfjFl1oMQiRVsFotDD/IMG_20150307_164437.jpg](https://s3.amazonaws.com/pushbullet-uploads/ujEbQh7xysS-PaQBCYlDl5pfMlLfjFl1oMQiRVsFotDD/IMG_20150307_164437.jpg)

The view controller has a retain count of 2 at that point, 1 for being referenced by the navigation controller, and one from the Java-side instantiation.

If we now pop the view controller, it's retain count becomes 1 and it is removed from `CUSTOM_OBJECTS`. However, it's still reachable via the lambda/Listener, which is referenced by the `ListenerWrapper`, which is referenced by both `CUSTOM_OBJECTS`as well as `ASSOCIATED_OBJECTS`.

The GC will never collect the view controller because of this, so all native and Java side objects stay alive.

If you manually remove the listener from the UIButton as outlined in my last comment, the cycle between lambda and view controller is broken, and the GC can collect all the things and release the native peers.

**Comment by mirkonasato** [comment]
Thanks for the thorough investigation. Happy to confirm that v1.0.0 fixes the issue with my real app (not just the sample) as long as listeners are removed as suggested.

Have to say having to manually remove listeners still feels cumbersome and error prone though, no much better than explicitly calling dispose(). Hope a better solution can be found at some point.

**Comment by badlogic** [comment]
Yes, not happy with manual listener removal either. We are working on it
and will leave this issue open. Thanks for reporting!

**Comment by mirkonasato** [comment]
Uhm, seems like I was using some 1.0.0-SNAPSHOT yesterday, not 1.0.0.

I actually still get a growing memory usage with 1.0.0. In fact 1.0.0-beta-04 works fine with the listener removal pattern, but not 1.0.0. I wonder if something else has changed that adds more references.

**Comment by badlogic** [comment]
The GC has to run to collect the controller and release the native peers. I
can't reproduce the issue with 1.0.0 or 1.0.1-SNAPSHOT when invoking the GC
manually. I can reproduce it when relying on the GC to decide when to
collect. If there's not enough memory pressure, then the GC will not do
it's thing. You can explicitly call System.gc which will most of the time
start a collection. You can also call ObjCObject.setLogRelease to enable
logging of custom object retain release calls.

**Comment by mirkonasato** [comment]
I was just trying ObjCObject.logRetainRelease(true), and the funny thing is that with that flag controllers do get collected.

I've updated the RoboLeaker sample with listener removal. Without logRetainRelease(true) I create 50 controllers and none get removed. With logRetainRelease(true) after creating 7 of them they do get removed.

Will try explicit GC...

**Comment by badlogic** [comment]
That's indeed weird. Let us know if explicit GC works for you. I'll take a
look on monday if the issue persists.

**Comment by mirkonasato** [comment]
Ok, calling System.gc() in didReceiveMemoryWarning() results in controllers being freed when triggering a memory warning in the simulator.

That's with the RoboLeaker sample, my real app is still not behaving... there must be something else holding references.
